File: Remote/Source/Singletons/DatabaseLoader.m
//
// DataBaseLoader.m
// Remote
//
// Created by Jason Cardwell on 2/19/12.
// Copyright (c) 2012 Moondeer Studios. All rights reserved.
//

@import UIKit;
@import CoreData;
@import CocoaLumberjack;
@import MoonKit;
#import "MSRemoteMacros.h"
#import "DatabaseLoader.h"
//#import "RemoteElement.h"
#import "RemoteController.h"
//#import "Remote.h"
#import "Activity.h"
#import "Remote-Swift.h"

#define USER_CODES_PLIST    @"UserCodes"
#define CODE_DATABASE_PLIST @"CodeDatabase-Pruned"

#define LOG_IMPORT_FILE      1
#define LOG_PARSED_FILE      2
#define LOG_RESULTING_OBJECT 4

#define PARSE_ONLY NO

#define REMOTECONTROLLER_LOG_FLAG 0
#define REMOTE_LOG_FLAG           0
#define PRESET_LOG_FLAG           0
#define IMAGES_LOG_FLAG           0
#define POWERCOMMANDS_LOG_FLAG    0
#define MANUFACTURERS_LOG_FLAG    0
#define COMPONENTDEVICES_LOG_FLAG 0
#define NETWORKDEVICES_LOG_FLAG   0
#define IRCODES_LOG_FLAG          0

static int       ddLogLevel       = LOG_LEVEL_DEBUG;
static const int msLogContext     = (LOG_CONTEXT_BUILDING | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
static const int importLogContext = (LOG_CONTEXT_IMPORT | LOG_CONTEXT_FILE);
#pragma unused(ddLogLevel, msLogContext)

void logImportFile(NSString * fileName, NSString * fileContent, int flag) {
  if ((flag & LOG_IMPORT_FILE) == LOG_IMPORT_FILE)
    MSLogCDebugInContext(importLogContext, @"%@.json:\n%@\n", fileName, fileContent);
}

void logParsedImportFile(NSString * fileName, id parsedObject, int flag) {
  if ((flag & LOG_PARSED_FILE) == LOG_PARSED_FILE)
    MSLogCDebugInContext(importLogContext,
                         @"object parsed from %@.json:\n%@\n",
                         fileName,
                         [MSJSONSerialization JSONFromObject:parsedObject]);
}

void logImportedObject(id importedObject, int flag) {
  if ((flag & LOG_RESULTING_OBJECT) == LOG_RESULTING_OBJECT)
    MSLogCDebugInContextIf((importedObject != nil),
                           importLogContext,
                           @"JSON from imported object(s):\n%@\n",
                           [MSJSONSerialization JSONFromObject:[importedObject JSONObject]]);
}

@implementation DatabaseLoader

/// loadData
/// @return BOOL
+ (BOOL)loadData {
  MSLogDebug(@"beginning data load...");

  [CoreDataManager backgroundSaveWithBlockAndWait:^(NSManagedObjectContext *context) {
    @autoreleasepool { [self loadPresets:context]; }
  }];

  [CoreDataManager backgroundSaveWithBlockAndWait:^(NSManagedObjectContext * context) {
    @autoreleasepool { [self loadImages:context]; }
  }];

  [CoreDataManager backgroundSaveWithBlockAndWait:^(NSManagedObjectContext * context) {
    @autoreleasepool { [self loadManufacturers:context]; }
  }];

  [CoreDataManager saveWithBlockAndWait:^(NSManagedObjectContext * context) {
    @autoreleasepool {
      [self loadComponentDevices:context];
      [self loadNetworkDevices:context];
    }
  }];

  [CoreDataManager backgroundSaveWithBlockAndWait:^(NSManagedObjectContext * context) {
    @autoreleasepool {
      [self loadRemoteController:context];
      [self loadActivities:context];
      [self loadRemotes:context];
    }
  }];

  MSLogDebug(@"data load complete");

  return YES;
}

/// loadRemotes:
/// @param context
+ (void)loadRemotes:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading remotes...");

  NSString * fileName = @"Remote_Demo";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];

  assert(fileContent);
  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, REMOTE_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, REMOTE_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * remotes = [Remote importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%lu remotes imported", (unsigned long)[remotes count]);

  error = nil;
  [context save:&error];
  MSHandleErrors(error);

  logImportedObject(remotes, REMOTE_LOG_FLAG);
}

/// loadRemoteController:
/// @param context
+ (void)loadRemoteController:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading remote controller...");

  NSString * fileName = @"RemoteController";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, REMOTECONTROLLER_LOG_FLAG);
  MSDictionary * importObject = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObject, REMOTECONTROLLER_LOG_FLAG);
  if (PARSE_ONLY) return;

  RemoteController * remoteController = [RemoteController importObjectFromData:importObject
                                                                       context:context];
  MSLogDebug(@"remote controller imported? %@", BOOLString((remoteController != nil)));

  logImportedObject(remoteController, REMOTECONTROLLER_LOG_FLAG);
}


/// loadPresets:
/// @param context
+ (void)loadPresets:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading presets...");

  NSString * fileName = @"Preset";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, PRESET_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingFile:filePath options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, PRESET_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * presets = [PresetCategory importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%@ presets imported", [presets valueForKeyPath:@"@sum.totalItemCount"]);

  error = nil;
  [context save:&error];
  MSHandleErrors(error);

  logImportedObject(presets, PRESET_LOG_FLAG);
}

/// loadActivities:
/// @param context
+ (void)loadActivities:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading activities...");

  NSString * fileName = @"Activity";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, REMOTE_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, REMOTE_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * activities = [Activity importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%lu activities imported", (unsigned long)[activities count]);

  error = nil;
  [context save:&error];
  MSHandleErrors(error);

  logImportedObject(activities, REMOTE_LOG_FLAG);
}

/// loadManufacturers:
/// @param context
+ (void)loadManufacturers:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading manufacturers...");

 #define MANUFACTURER_TEST_CODES
  #ifdef MANUFACTURER_TEST_CODES
    NSString * fileName = @"Manufacturer_Test";
  #else
    NSString * fileName = @"Manufacturer";
  #endif
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, MANUFACTURERS_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingFile:filePath options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, MANUFACTURERS_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * manufacturers = [Manufacturer importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%lu manufacturers imported", (unsigned long)[manufacturers count]);

  logImportedObject(manufacturers, MANUFACTURERS_LOG_FLAG);
}

/// loadComponentDevices:
/// @param context
+ (void)loadComponentDevices:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading component devices...");

  NSString * fileName = @"ComponentDevice";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, COMPONENTDEVICES_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, COMPONENTDEVICES_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * componentDevices = [ComponentDevice importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%lu component devices imported", (unsigned long)[componentDevices count]);

  logImportedObject(componentDevices, COMPONENTDEVICES_LOG_FLAG);
}

/// loadNetworkDevices:
/// @param context
+ (void)loadNetworkDevices:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading network devices...");

  NSString * fileName = @"NetworkDevice";
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, NETWORKDEVICES_LOG_FLAG);
  NSArray * importObjects = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];

  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, NETWORKDEVICES_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * networkDevices = [NetworkDevice importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%lu network devices imported", (unsigned long)[networkDevices count]);

  logImportedObject(networkDevices, NETWORKDEVICES_LOG_FLAG);
}

/// loadImages:
/// @param context
+ (void)loadImages:(NSManagedObjectContext *)context {
  MSLogDebug(@"loading images...");
//  #define GLYPHISH_TEST_IMAGES
  #ifdef GLYPHISH_TEST_IMAGES
    NSString * fileName = @"Glyphish_Test";
  #else
    NSString * fileName = @"Glyphish";
  #endif
  NSString * filePath = [MainBundle pathForResource:fileName ofType:@"json"];

  NSError        * error = nil;
  NSStringEncoding encoding;
  NSString       * fileContent = [NSString stringWithContentsOfFile:filePath
                                                       usedEncoding:&encoding
                                                              error:&error];
  assert(fileContent);

  if (MSHandleErrors(error)) return;

  logImportFile(fileName, fileContent, IMAGES_LOG_FLAG);

  NSArray * importObjects = [JSONSerialization objectByParsingString:fileContent options:1 error:&error];
  if (MSHandleErrors(error)) return;

  logParsedImportFile(fileName, importObjects, IMAGES_LOG_FLAG);
  if (PARSE_ONLY) return;

  NSArray * images = [ImageCategory importObjectsFromData:importObjects context:context];
  MSLogDebug(@"%@ images imported", [images valueForKeyPath:@"@sum.totalItemCount"]);

  logImportedObject(images, IMAGES_LOG_FLAG);
}

@end

File: Remote/Source/Singletons/DatabaseLoader.swift
//
//  DatabaseLoader.swift
//  Remote
//
//  Created by Jason Cardwell on 12/22/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc class DatabaseLoader {

  /** loadData */
  class func loadData(completion: ((Bool, NSError?) -> Void)? = nil) {

    let moc = DataManager.rootContext

    moc.performBlock {
      self.loadPresets(moc)
      self.loadImages(moc)
      self.loadManufacturers(moc)
      self.loadComponentDevices(moc)
      self.loadActivityController(moc)
      self.loadActivities(moc)
      self.loadRemotes(moc)
      var error: NSError?
      MSLogDebug("saving context…")
      let success = moc.save(&error)
      if success { MSLogDebug("context saved successfully") }
      else { MSLogDebug("context failed to save") }
      completion?(success, error)
    }

/*    DataManager.saveContext(DataManager.rootContext, withBlock: {
        self.loadPresets($0)
        self.loadImages($0)
        self.loadManufacturers($0)
        self.loadComponentDevices($0)
//        self.loadNetworkDevices($0)
        self.loadActivityController($0)
        self.loadActivities($0)
        self.loadRemotes($0)
      },
      nonBlocking: true,
      completion: completion)	*/
  }

  /**
  loadRemotes:

  :param: context NSManagedObjectContext
  */
  private class func loadRemotes(context: NSManagedObjectContext) {
    MSLogDebug("loading remotes...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("Remote_Demo", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = Remote.importObjectsFromData(importData, context: context)
      MSLogDebug("\(importedObjects.count ?? 0) remotes imported")
    }
  }

  /**
  loadActivityController:

  :param: context NSManagedObjectContext
  */
  private class func loadActivityController(context: NSManagedObjectContext) {
    MSLogDebug("loading activity controller...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("ActivityController", ofType: "json") {
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSDictionary
      assert(MSHandleError(error) == false)
      let importedObject = ActivityController.importObjectFromData(importData as! [NSObject:AnyObject], context: context)
      MSLogDebug("activity controller imported? \(importedObject != nil)")
    }
  }


  /**
  loadPresets:

  :param: context NSManagedObjectContext
  */
  private class func loadPresets(context: NSManagedObjectContext) {
    MSLogDebug("loading presets...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("Preset", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = PresetCategory.importObjectsFromData(importData, context: context) as? [PresetCategory]
      MSLogDebug("\(reduce(importedObjects?.map({$0.totalItemCount}) ?? [], 0, {$0 + $1})) presets imported")
    }
  }

  /**
  loadActivities:

  :param: context NSManagedObjectContext
  */
  private class func loadActivities(context: NSManagedObjectContext) {
    MSLogDebug("loading activities...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("Activity", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = Activity.importObjectsFromData(importData, context: context)
      MSLogDebug("\(importedObjects.count ?? 0) activities imported")
    }
  }

  /**
  loadManufacturers:

  :param: context NSManagedObjectContext
  */
  private class func loadManufacturers(context: NSManagedObjectContext) {
    MSLogDebug("loading manufacturers...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("Manufacturer_Test", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = Manufacturer.importObjectsFromData(importData, context: context)
      MSLogDebug("\(importedObjects.count ?? 0) manufacturers imported")
    }
  }

  /**
  loadComponentDevices:

  :param: context NSManagedObjectContext
  */
  private class func loadComponentDevices(context: NSManagedObjectContext) {
    MSLogDebug("loading component devices...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("ComponentDevice", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = ComponentDevice.importObjectsFromData(importData, context: context)
      MSLogDebug("\(importedObjects.count ?? 0) component devices imported")
    }
  }

  /**
  loadNetworkDevices:

  :param: context NSManagedObjectContext
  */
  private class func loadNetworkDevices(context: NSManagedObjectContext) {
    MSLogDebug("loading network devices...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("NetworkDevice", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = NetworkDevice.importObjectsFromData(importData, context: context)
      MSLogDebug("\(importedObjects.count ?? 0) network devices imported")
    }
  }

  /**
  loadImages:

  :param: context NSManagedObjectContext
  */
  private class func loadImages(context: NSManagedObjectContext) {
    MSLogDebug("loading images...")
    var error: NSError?
    if let filePath = NSBundle.mainBundle().pathForResource("Glyphish", ofType: "json"),
      let importData = JSONSerialization.objectByParsingFile(filePath, options:1, error:&error) as? NSArray {
      assert(MSHandleError(error) == false)
      let importedObjects = ImageCategory.importObjectsFromData(importData, context: context) as! [ImageCategory]
      MSLogDebug("\(reduce(importedObjects.map({$0.totalItemCount}) ?? [], 0, {$0 + $1})) images imported")
    }
  }


}
